crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -1659,78 +1659,21 @@ pub(crate) fn validate_cursor_pagination(
+    let report_records: Option<BTreeMap<NativeSuiAddress, Vec<NativeSuiAddress>>> =
+        system_state.map(|x| BTreeMap::from_iter(x.validator_report_records));
 
-            let credentials = ValidatorCredentials {
-                protocol_pub_key: Some(Base64::from(v.protocol_pubkey_bytes.clone())),
-                network_pub_key: Some(Base64::from(v.network_pubkey_bytes.clone())),
-                worker_pub_key: Some(Base64::from(v.worker_pubkey_bytes.clone())),
-                proof_of_possession: Some(Base64::from(v.proof_of_possession_bytes.clone())),
-                net_address: Some(v.net_address.clone()),
-                p2p_address: Some(v.p2p_address.clone()),
-                primary_address: Some(v.primary_address.clone()),
-                worker_address: Some(v.worker_address.clone()),
-            };
-            Validator {
-                address: Address {
-                    address: SuiAddress::from(v.sui_address),
-                },
-                next_epoch_credentials: Some(credentials.clone()),
-                credentials: Some(credentials),
-                name: Some(v.name.clone()),
-                description: Some(v.description.clone()),
-                image_url: Some(v.image_url.clone()),
-                project_url: Some(v.project_url.clone()),
-
-                operation_cap_id: SuiAddress::from_array(**v.operation_cap_id),
-                staking_pool_id: SuiAddress::from_array(**v.staking_pool_id),
-                exchange_rates_id: SuiAddress::from_array(**v.exchange_rates_id),
-                exchange_rates_size: Some(v.exchange_rates_size),
-
-                staking_pool_activation_epoch: v.staking_pool_activation_epoch,
-                staking_pool_sui_balance: Some(BigInt::from(v.staking_pool_sui_balance)),
-                rewards_pool: Some(BigInt::from(v.rewards_pool)),
-                pool_token_balance: Some(BigInt::from(v.pool_token_balance)),
-                pending_stake: Some(BigInt::from(v.pending_stake)),
-                pending_total_sui_withdraw: Some(BigInt::from(v.pending_total_sui_withdraw)),
-                pending_pool_token_withdraw: Some(BigInt::from(v.pending_pool_token_withdraw)),
-                voting_power: Some(v.voting_power),
-                // stake_units: todo!(),
-                gas_price: Some(BigInt::from(v.gas_price)),
-                commission_rate: Some(v.commission_rate),
-                next_epoch_stake: Some(BigInt::from(v.next_epoch_stake)),
-                next_epoch_gas_price: Some(BigInt::from(v.next_epoch_gas_price)),
-                next_epoch_commission_rate: Some(v.next_epoch_commission_rate),
-                at_risk,
-                report_records,
-                // apy: todo!(),
-            }
+    validators
+        .iter()
+        .map(|v| Validator {
+            validator_summary: v.clone(),
+            at_risk_validators: at_risk_validators.clone(),
+            report_records: report_records.clone(),
```

Comment:
You're most of the way there with this, just do the lookup into each of these maps to produce an `Option<u64>` and an `Option<Vec<SuiAddress>>` to pass in here, rather than copying the whole map into each validator to fetch one key out of it later on.